python/interpret_community/mimic/models/lightgbm_model.py

```diff
@@ -59,6 +62,69 @@ def predict_wrapper(self, X, *args, **kwargs):
         return self._function(X, *args, **kwargs)
 
 
+class _SparseTreeExplainer(object):
+
+    """Wraps the lightgbm model to enable sparse feature contributions.
+
+    If version is >=3.1.0, runs on sparse input data by calling predict function directly.
+
+    :param lgbm: The lightgbm model to wrap.
+    :type lgbm: LGBMModel
+    :param tree_explainer: The tree_explainer used for dense data.
+    :type tree_explainer: shap.TreeExplainer
+    """
+
+    def __init__(self, lgbm, tree_explainer):
+        """Wraps the lightgbm model to enable sparse feature contributions.
+
+        :param lgbm: The lightgbm model to wrap.
+        :type lgbm: LGBMModel
+        :param tree_explainer: The tree_explainer used for dense data.
+        :type tree_explainer: shap.TreeExplainer
+        """
+        self._lgbm = lgbm
+        self._tree_explainer = tree_explainer
+        self._num_iters = -1
+        # Get the number of iterations trained for from the booster
+        if hasattr(self._lgbm._Booster, 'params'):
+            if NUM_ITERATIONS in self._lgbm._Booster.params:
+                self._num_iters = self._lgbm._Booster.params[NUM_ITERATIONS]
+        # If best iteration specified, use that
+        if self._lgbm._best_iteration is not None:
+            self._num_iters = self._lgbm._best_iteration
+        self.expected_value = None
+
+    def shap_values(self, X):
+        """Calls lightgbm predict directly for sparse case.
+
+        If lightgbm version is >=3.1.0, runs on sparse input data
+        by calling predict function directly with pred_contrib=True.
+        Uses tree explainer for dense input data.
+
+        :param X: The model evaluation examples.
+        :type X: numpy.array or scipy.sparse.csr_matrix
+        :return: The feature importance values.
+        :rtype: numpy.array, scipy.sparse or list of scipy.sparse
+        """
+        if issparse(X):
+            shap_values = self._lgbm.predict(X,
+                                             num_iteration=self._num_iters,
+                                             pred_contrib=True)
+            if isinstance(shap_values, list):
+                shape = shap_values[0].shape
+                self.expected_value = shap_values[0][0, shape[1]-1]
+                for idx, class_values in enumerate(shap_values):
+                    shap_values[idx] = class_values[:, :shape[1]-1]
```

Comment:
Nit: spaces around `-` operator (also see above and below), I'm surprised linting builds didn't catch this?

Reply:
interesting, running:

```
flake8 --max-line-length=119
```

seems to pass either way.  Anyway I've updated it to have spaces, thanks!